doc/source/dev/contributor/using_blas_lapack.rst

```diff
@@ -0,0 +1,213 @@
+
+In existing C/C++ code, there are several different approaches, usually
+motivated by the same symbol names and ABI issues that we discussed higher up
+on this page. We give two examples here, for SuperLU and for ``trlib``.
+
+In SuperLU (part of ``sparse/linalg/_dsolve/``), symbol names used contain only
+a trailing underscore, for example ``zgemv_``. This canonical name is mapped
+within ``SuperLU/SRC/slu_Cnames.h`` to a host of other names depending on some
+build config. The correct one is selected for SciPy in
+``SuperLU/SRC/scipy_slu_config.h``.
+
+In ``optimize/_trlib``, the C code uses names prefixed with ``TRLIB_``. These names are created
+in ``optimize/_trlib/trlib_private.h`` - for example for ``TRLIB_DAXPY`` we have:
+
+.. code:: c
+
+    /* in `scipy/_build_utils/src/`, provides `BLAS_FUNC` and `CBLAS_INT` */
+    #include "npy_cblas.h"
+
+    void BLAS_FUNC(daxpy)(CBLAS_INT *n, double *alpha, double *x, CBLAS_INT *incx, double *y, CBLAS_INT *incy);
+
+    static void trlib_daxpy(trlib_int_t *n, double *alpha, double *x, trlib_int_t *incx, double *y, trlib_int_t *incy)
+    {
+        CBLAS_INT n_ = *n, incx_ = *incx, incy_ = *incy;
+        BLAS_FUNC(daxpy)(&n_, alpha, x, &incx_, y, &incy_);
+    }
+
+    /*
+     * The lower-case to upper-case `TRLIB_DAXPY` is done with an ugly macro
+     * which also uses timer functions to allow benchmarking - we leave that out
+     * here (just don't do it like that!)
+     */
+
+When we're vendoring C or C++ code that is already using BLAS or LAPACK, we
+typically see that that code has already figured out some way of naming their
+functions (a la ``trlib_daxpy``), and we may have to adapt to it (or fix it
+up).
+
+That leaves the question what we should do when we're writing new C or C++
+code. The key ingredients are ``BLAS_FUNC`` for name mangling and ``CBLAS_INT``
+for LP64/ILP64 support, both defined in ``scipy/_build_utils/src/npy_cblas.h``.
+
+**TODO finish describing canonical method here, and link to example! :**
+
+- Practical example to fix up: https://github.com/scipy/scipy/pull/19970
+  (interpolate refactor, using ``dlartg``)
+- Note: CBLAS functions are already provided by the ``npy_cblas.h`` header as
+  well, and NumPy uses those.
+- We could use CBLAS functions in SciPy too, but don't do that as of now. Given
+  that we shouldn't rely on LAPACKE, it doesn't quite make sense to rely on
+  CBLAS either probably. Rather, we should use the Fortran symbols for both
+  BLAS and LAPACK directly.
+- To use a BLAS or LAPACK function:
+
+  - use ``#include "scipy_lapack.h"`` (which includes ``npy_cblas.h"``)
+  - if the routine you need isn't present yet, add it in ``scipy_lapack.h``.
+    TBD: do we generate this header - it's basically
+    ``build/scipy/linalg/_lapack_subroutines.h`` with ``C_INT`` instead of
+    ``int``...
+  - call the functions as ``BLAS_FUNC(name)``, where ``name`` is the canonical
+    name for the BLAS or LAPACK function (e.g., ``daxpy``)
```

Comment:
What seems to work today is
```diff
$ git diff HEAD^
diff --git a/scipy/interpolate/meson.build b/scipy/interpolate/meson.build
index 4c3631c0e1..25973a0846 100644
--- a/scipy/interpolate/meson.build
+++ b/scipy/interpolate/meson.build
@@ -123,8 +123,7 @@ py3.extension_module('_rgi_cython',
 
 __fitpack_lib = static_library('__fitpack',
     ['src/__fitpack.h', 'src/__fitpack.cc'],
-    dependencies:[lapack]
+    dependencies:[lapack, np_dep, py3_dep]
 )
 
 __fitpack_dep = declare_dependency(
diff --git a/scipy/interpolate/src/__fitpack.h b/scipy/interpolate/src/__fitpack.h
index 471eca55a8..6f71adb7a2 100644
--- a/scipy/interpolate/src/__fitpack.h
+++ b/scipy/interpolate/src/__fitpack.h
@@ -2,9 +2,11 @@
 #include <iostream>
 #include <tuple>
 #include <vector>
+
+#include "../_build_utils/src/npy_cblas.h"
 #include "../_build_utils/src/fortran_defs.h"
 
-#define DLARTG F_FUNC(dlartg, DLARTG)
+#define DLARTG BLAS_FUNC(dlartg)

```

It'd be great indeed to generate the header file so that it can be included --- unless the dance for telling meson to use a generated header is less than this (which I'm not sure about TBH).